### ISSUE TITLE
Peer Review Lab02

### ISSUE THREAD

**Issue by lucabubi** [Farisanfekri/Computational-Intelligence#5]
Hello there 👋,
First of all I wanna thank you for sharing your code with us.
You've added a reasonable amount of comments to the code, which is a good thing.
The algorithm you provided seems fine but I would suggest you some changes regarding different aspects of the code.

I suggest changing the line
```
NUM_MOVES = sum([i * 2 + 1 for i in range(NUM_ROWS)])//2
```
to
```
MAX_NUM_MOVES = sum([i * 2 + 1 for i in range(NUM_ROWS)])//2
```
for better understanding. This reflects a limit-case because it's not typical for both your algorithm and the opponent to remove only one object per turn in each round.

Regarding the _**fitness**_ function, I geniunely appriciate the change of strategy stated by
```
ply = strategy[i%3](nim)
```
because that ensures a rotational pattern for the oppent's strategy selection.
I also like a lot this part
```
if player == 0:
 score = score + 10
else:
 score = score + num_moves*0.01
```
in which you seem to give importance to the number of moves. A higher number of moves may contribute to a higher score, reflecting a preference not only for winning strategies but also for strategies that involve more moves in a game.
I suggest you to change
```
for i in range(18):
```
maybe into
```
NUM_ROUNDS = 18
for i in range(NUM_ROUNDS)
```
alwayss for better comprehension.
I also would have suggested you to increase this number but it would be probabily useless because I see a "main problem" I think you should take into consideration:
I notice that your algorithm's moves are determined **prior to the beginning of the game**, as stated by the following code
```
for _ in range(NUM_MOVES):
 row = randint(0, NUM_ROWS-1)
 i.genotype.extend([(row, min(randint(1, row * 2 + 1), UPPERBOUND_K))])
```
without considering the current state of the game or the moves executed by your opponent. These moves are not adjusted based on these parameters during each turn.
So, basically, your algorithm just randomly makes moves without considering what's happening in the game or what the opponent is doing.
With that said, I think you've done a good job, although there's room for improvement.
Good luck for the next labs! 👋
